viewer/packages/camera-manager/src/CameraManager.ts

```diff
@@ -0,0 +1,534 @@
+   * @param cursorPosition.y
+   */
+  private calculateMissedRaycast(
+    cursorPosition: { x: number; y: number },
+    modelsBoundingBox: THREE.Box3
+  ): THREE.Vector3 {
+    const modelSize = modelsBoundingBox.min.distanceTo(modelsBoundingBox.max);
+
+    this._raycaster.setFromCamera(cursorPosition, this._camera);
+
+    const farPoint = this._raycaster.ray.direction
+      .clone()
+      .normalize()
+      .multiplyScalar(
+        Math.max(this._camera.position.distanceTo(modelsBoundingBox.getCenter(new THREE.Vector3())), modelSize)
+      )
+      .add(this._camera.position);
+
+    return farPoint;
+  }
+
+  /**
+   * Changes controls target based on current cursor position.
+   * @param event MouseEvent that contains pointer location data.
+   */
+  private async changeTarget(event: MouseEvent) {
+    const { offsetX, offsetY } = event;
+
+    const { x, y } = this.convertPixelCoordinatesToNormalized(offsetX, offsetY);
+
+    const callbackData = await this._modelRaycastCallback(offsetX, offsetY);
+
+    const newTarget =
+      callbackData?.intersection?.point ?? this.calculateMissedRaycast({ x, y }, callbackData.modelsBoundingBox);
+
+    this.setCameraTarget(newTarget, true);
+  }
+
+  /**
+   * Changes controls scroll target based on current cursor position.
+   * @param event MouseEvent that contains pointer location data.
+   */
+  private async changeScrollTarget(event: any) {
+    const { offsetX, offsetY } = event;
+    const { x, y } = this.convertPixelCoordinatesToNormalized(offsetX, offsetY);
+
+    const callbackData = await this._modelRaycastCallback(offsetX, offsetY);
+
+    const newScrollTarget =
+      callbackData?.intersection?.point ?? this.calculateMissedRaycast({ x, y }, callbackData.modelsBoundingBox);
+
+    this.controls.setScrollTarget(newScrollTarget);
+  }
+
+  /**
+   * Removes controls event listeners if they are defined.
+   */
+  private teardownControls() {
+    if (this._onClick !== undefined) {
+      this._domElement.removeEventListener('click', this._onClick);
+      this._onClick = undefined;
+    }
+    if (this._onWheel !== undefined) {
+      this._domElement.removeEventListener('wheel', this._onWheel);
+      this._onWheel = undefined;
+    }
+  }
+
+  /**
+   * Method for setting up camera controls listeners and values inside current controls class.
+   */
+  private setupControls() {
+    let startedScroll = false,
+      newTargetUpdate = false;
+    let timeAfterClick = 0;
+    const wheelClock = new THREE.Clock(),
+      clickClock = new THREE.Clock();
+
+    const onClick = (e: any) => {
+      newTargetUpdate = true;
+      timeAfterClick = 0;
+      clickClock.getDelta();
+
+      this.controls.enableKeyboardNavigation = false;
+      this.changeTarget(e);
+    };
+
+    const onWheel = async (e: any) => {
+      const timeDelta = wheelClock.getDelta();
+      timeAfterClick += clickClock.getDelta();
+
+      if (timeAfterClick > 3) newTargetUpdate = false;
```

Comment:
What is "3"? Seconsd?

Reply:
Yes, but not sure about that mechanics and how it feels. It is preventing zoom to cursor functionality for 3 seconds after target change click was initiated. Do you think it's bad idea or we can keep it and add other parameter for it?

Reply:
Hmm, it sounds a bit odd and that it easily could lead to confusion.

Reply:
Removed, finally